src/context.rs

```diff
@@ -10,6 +10,211 @@ use crate::ffi::types::{c_uint, c_void, AlignedType};
+
+    const MAX_SPINLOCK_ATTEMPTS: usize = 128;
+    const MAX_PREALLOC_SIZE: usize = 16; // measured at 208 bytes on Andrew's 64-bit system
+
+    static SECP256K1: SpinLock = SpinLock::new();
+
+    // Simple spinlock-gated structure which holds the backing store for a
+    // secp256k1 context.
+    //
+    // To obtain exclusive access, call [`Self::try_lock`], which will spinlock
+    // for some small number of iterations before giving up. By trying again in
+    // a loop, you can emulate a "true" spinlock that will only yield once it
+    // has access. However, this would be very dangerous, especially in a nostd
+    // environment, because if we are pre-empted by an interrupt handler while
+    // the lock is held, and that interrupt handler attempts to take the lock,
+    // then we deadlock.
+    //
+    // Instead, the strategy we take within this module is to simply create a
+    // new stack-local context object if we are unable to obtain a lock on the
+    // global one. This is slow and loses the defense-in-depth "rerandomization"
+    // anti-sidechannel measure, but it is better than deadlocking..
+    struct SpinLock {
+        flag: AtomicBool,
+        // Invariant: if this is non-None, then the store is valid and can be
+        // used with `ffi::secp256k1_context_preallocated_create`.
+        data: UnsafeCell<([AlignedType; MAX_PREALLOC_SIZE], Option<NonNull<ffi::Context>>)>,
+    }
+
+    // Required by rustc if we have a static of this type.
+    // Safety: `data` is accessed only while the `flag` is held.
+    unsafe impl Sync for SpinLock {}
+    unsafe impl Send for SpinLock {}
+
+    impl SpinLock {
+        const fn new() -> Self {
+            Self {
+                flag: AtomicBool::new(false),
+                data: UnsafeCell::new(([AlignedType::ZERO; MAX_PREALLOC_SIZE], None)),
+            }
+        }
+
+        /// Blocks until the lock is acquired, then returns an RAII guard.
+        fn try_lock(&self) -> Option<SpinLockGuard<'_>> {
+            for _ in 0..MAX_SPINLOCK_ATTEMPTS {
+                // `compare_exchange_weak` is fine here: we’re spinning anyway.
+                if self
+                    .flag
+                    .compare_exchange_weak(false, true, Ordering::Acquire, Ordering::Relaxed)
+                    .is_ok()
+                {
+                    return Some(SpinLockGuard { lock: self });
+                }
+                spin_loop();
+            }
+            None
+        }
+
+        #[inline(always)]
+        fn unlock(&self) { self.flag.store(false, Ordering::Release); }
+    }
+
+    /// Drops the lock when it goes out of scope.
+    pub struct SpinLockGuard<'a> {
+        lock: &'a SpinLock,
+    }
+
+    impl Deref for SpinLockGuard<'_> {
+        type Target = ([AlignedType; MAX_PREALLOC_SIZE], Option<NonNull<ffi::Context>>);
+        fn deref(&self) -> &Self::Target {
+            // Safe: we hold the lock.
```

Comment:
For inspiration: I've started to use the word "sound" everywhere instead of "safe" since it's more correct. (The block literally says "unsafe" so it cannot be safe but it can be sound"). I definitely don't require this but if you feel like it makes sense you might replace the word. :)